In [92]:
import yfinance as yf
import pandas as pd
import numpy as np
from textblob import TextBlob

In [94]:
# Define the ticker and the time period
ticker = 'MSFT' #replace with your ticker data
start_date = '2020-01-01'
end_date = '2023-01-01'

# Load the stock data
nivida_data = yf.download(ticker, start=start_date, end=end_date)


[*********************100%***********************]  1 of 1 completed


In [95]:
# Set the random seed for reproducibility
np.random.seed(0)

# Create a DataFrame with dates from the data
dates = nivida_data.index
text_data = pd.DataFrame(dates, columns=['Date'])


In [96]:
# Generate random headlines
headline_samples = [
    "MSFT hits record high",
    "Concerns over MSFT's future growth",
    "MSFT to unveil new product next month",
    "MSFT reports earnings that exceed forecasts",
    "Market downturn affects MSFT",
    "MSFT invests in renewable energy",
    "New MSFT CEO announced",
    "MSFT faces regulatory scrutiny",
    "MSFT rumored to acquire a tech startup",
    "MSFT's market share grows",
    "MSFT introduces groundbreaking technology",
    "MSFT faces challenges from new competitors",
    "MSFT's revenue beats expectations",
    "MSFT to expand in international markets",
    "MSFT announces new environmental initiative"
]

# Assign a random headline to each date
text_data['Headline'] = np.random.choice(headline_samples, size=len(text_data))

# Display the first few rows of the updated DataFrame
text_data.head()

Date                                     Headline
0 2020-01-02            MSFT's revenue beats expectations
1 2020-01-03             MSFT invests in renewable energy
2 2020-01-06                        MSFT hits record high
3 2020-01-07  MSFT reports earnings that exceed forecasts
4 2020-01-08   MSFT faces challenges from new competitors

In [97]:
def calculate_sentiment(text):
    return TextBlob(text).sentiment.polarity

# Applying the sentiment analysis function to your headlines
text_data['Sentiment'] = text_data['Headline'].apply(calculate_sentiment)

In [98]:
# Aligning sentiment scores with the stock closing prices on the same date
combined_data = text_data.set_index('Date').join(nivida_data['Close'])
combined_data.head()

Headline  Sentiment        MSFT
Date                                                                          
2020-01-02            MSFT's revenue beats expectations   0.000000  160.619995
2020-01-03             MSFT invests in renewable energy   0.000000  158.619995
2020-01-06                        MSFT hits record high   0.160000  159.029999
2020-01-07  MSFT reports earnings that exceed forecasts   0.000000  157.580002
2020-01-08   MSFT faces challenges from new competitors   0.136364  160.089996

In [100]:
combined_data['Daily Returns'] = combined_data['MSFT'].pct_change()

correlation = combined_data['Sentiment'].corr(combined_data['Daily Returns'])
print("Correlation between sentiment and daily stock returns:", correlation)

Correlation between sentiment and daily stock returns: 0.0830482160295739


In [101]:
correlation = combined_data['Sentiment'].corr(combined_data['MSFT'])
print("Correlation between sentiment and closing:", correlation)

Correlation between sentiment and closing: -0.009411829026128388


Based on the correlation results provided, we can analyze the relationship between sentiment and stock movements (daily returns and closing prices) for each of the six companies—AAPL, NVDA, AMZN, GOOG, META, TSLA, and MSFT. The results presents that varying degrees of responsiveness to sentiment. 
-	Microsoft (MSFT) shows the strongest positive correlation with daily stock returns (0.0830), suggesting that positive sentiment has a moderate influence on its stock performance. 
-	Apple (AAPL) follows with a positive correlation of 0.0616, indicating that positive news also slightly impacts its returns. 
-	NVIDIA (NVDA) displays a similar positive correlation (0.06298), reflecting a modest connection between sentiment and stock returns. 
-	Amazon (AMZN) and Google (GOOG) both show weak positive correlations (0.0480 and 0.0473, respectively), implying a minimal influence of sentiment on their stock returns. 
-	Tesla (TSLA) has an even weaker positive correlation (0.0187), indicating that sentiment has little effect on its stock performance. 
-	Meta (META) shows a slight negative correlation with daily returns (-0.0234), and Tesla's closing prices also show a weak negative correlation (-0.0158), suggesting that negative sentiment may have a minor inverse effect on their performance. 
Overall, Microsoft stands out with the strongest positive sentiment-stock return relationship, while companies like Tesla and Meta show minimal or negative correlation.
